File: backend/core/src/modules/world-settings/world-settings.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { ProjectAccessService } from '../../common/access/project-access.service';
import { ProjectScopedCrudService } from '../../common/access/project-scoped-crud.service';
import { WorldSetting } from '../../entities';
import { SearchService } from '../search/search.service';

@Injectable()
export class WorldSettingsService extends ProjectScopedCrudService<WorldSetting> {
  constructor(
    @InjectRepository(WorldSetting)
    repo: Repository<WorldSetting>,
    private readonly searchService: SearchService,
    projectAccessService: ProjectAccessService,
  ) {
    super(repo, projectAccessService);
  }

  override async createForUser(userId: string | undefined, dto: Partial<WorldSetting>): Promise<WorldSetting> {
    const setting = await super.createForUser(userId, dto);
    await this.searchService.indexDocument('world_settings', setting.id, {
      id: setting.id,
      projectId: setting.projectId,
      title: setting.title,
      content: setting.content,
      category: setting.category,
    });
    return setting;
  }

  override async updateForUser(userId: string | undefined, id: string, dto: Partial<WorldSetting>): Promise<WorldSetting | null> {
    const setting = await super.updateForUser(userId, id, dto);
    if (setting) {
      await this.searchService.indexDocument('world_settings', setting.id, {
        id: setting.id,
        projectId: setting.projectId,
        title: setting.title,
        content: setting.content,
        category: setting.category,
      });
    }
    return setting;
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/service/WorldSettingsService.java
package store.pjcloud.cowrite.core.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.Map;
import java.util.UUID;
import org.springframework.stereotype.Service;
import store.pjcloud.cowrite.core.common.ProjectScopedCrudService;
import store.pjcloud.cowrite.core.entity.WorldSetting;
import store.pjcloud.cowrite.core.repository.WorldSettingRepository;

@Service
public class WorldSettingsService extends ProjectScopedCrudService<WorldSetting> {
    private final SearchService searchService;

    public WorldSettingsService(WorldSettingRepository repository,
                                ProjectAccessService projectAccessService,
                                ObjectMapper objectMapper,
                                SearchService searchService) {
        super(repository, projectAccessService, objectMapper, WorldSetting.class);
        this.searchService = searchService;
    }

    @Override
    public WorldSetting createForUser(UUID userId, Map<String, Object> body) {
        WorldSetting setting = super.createForUser(userId, body);
        index(setting);
        return setting;
    }

    @Override
    public WorldSetting updateForUser(UUID userId, UUID id, Map<String, Object> body) {
        WorldSetting setting = super.updateForUser(userId, id, body);
        if (setting != null) index(setting);
        return setting;
    }

    private void index(WorldSetting setting) {
        Map<String, Object> doc = new java.util.HashMap<>();
        doc.put("id", setting.getId() == null ? null : setting.getId().toString());
        doc.put("projectId", setting.getProjectId() == null ? null : setting.getProjectId().toString());
        doc.put("title", setting.getTitle());
        doc.put("content", setting.getContent());
        doc.put("category", setting.getCategory());
        searchService.indexDocument("world_settings", setting.getId().toString(), doc);
    }
}
